Composer/packages/server/src/server.ts

```diff
@@ -19,115 +19,118 @@ import { apiRouter } from './router/api';
 import { BASEURL } from './constants';
 import { attachLSPServer } from './utility/attachLSP';
 import log from './logger';
+import pluginLoader from './services/pluginLoader';
 
 const app: Express = express();
 app.set('view engine', 'ejs');
 app.set('view options', { delimiter: '?' });
 app.use(compression());
 
-const { login, authorize } = getAuthProvider();
-
-const CS_POLICIES = [
-  "default-src 'none';",
-  "font-src 'self' https:;",
-  "img-src 'self' data:;",
-  "base-uri 'none';",
-  "connect-src 'self';",
-  "frame-src 'self' bfemulator: https://login.microsoftonline.com https://*.botframework.com;",
-  "worker-src 'self';",
-  "form-action 'none';",
-  "frame-ancestors 'self';",
-  "manifest-src 'self';",
-  'upgrade-insecure-requests;',
-];
-
-app.all('*', function(req: Request, res: Response, next: NextFunction) {
-  res.header('Access-Control-Allow-Origin', '*');
-  res.header('Access-Control-Allow-Methods', 'GET,HEAD,OPTIONS,POST,PUT,DELETE');
-  res.header('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept, Authorization');
-
-  if (process.env.ENABLE_CSP === 'true') {
-    req.__nonce__ = crypto.randomBytes(16).toString('base64');
-    res.header(
-      'Content-Security-Policy',
-      CS_POLICIES.concat([
-        `script-src 'self' 'nonce-${req.__nonce__}';`,
-        // TODO: use nonce strategy after addressing issues with monaco-editor pacakge
-        "style-src 'self' 'unsafe-inline'",
-        // `style-src 'self' 'nonce-${req.__nonce__}';`,
-      ]).join(' ')
-    );
-  }
+pluginLoader.loadPluginsFromFolder(__dirname + '/../../../plugins', app).then(() => {
```

Comment:
loadPluginsFromFolders is great.  Can you support also loading from code, by that, i was facing such a scenario:

I was trying to create a HttpPublisherGenerator which takes a config (name, host, etc) and return a IPublisher, so that for different slot, i will register different Publisher hopefully like this 
```typescript 
var httpPublishers:  IPublisher = [
    HttpPublisherGenerator(config1), // config 1 with product slot url
    HttpPublisherGenerator(config2), // config 2 with testing slot url  
]
```

Reply:
Good call!

Reply:
Having seen your implementation, I don't 100% agree with this.

I think you would register ONE instance of the HttpPublisher, and then pass in the appropriate config, which is a bot-specific setting rather than a Composer-wide setting.

Reply:
Thanks, saw you comments there, i haven't thought about whether the publisher configuration is per-bot or per-composer, my gut feeling is making the config per bot would be more flexible, on the other hand making the configuration in composer will be easier to manage and a little clean.  Let me mark that as a open question and think about that. cc 
two chris @christopheranderson  @cwhitten 

One of my focus on prototype was trying to setup a structure that code-based publisher and config-based plugin (throw HttpPublisher) can work well together. So that, average users don't have to code to enable a plugin (i believe that's part of intention of your original design doc) and we still keep the possibility to let user code when user want to or have to, for features that can't achieve throw by configuring HttpPublisher, which meet @christopheranderson's point at that meeting.

Anyway, either way it's OK to me, and we need to make sure we surface the configuration out, become multiple plugins or multiple profiles (for one plugin).